src/Layout.js

```diff
@@ -1431,35 +1431,70 @@ var computeLayout = (function() {
-        cachedLayout.availableHeight === availableHeight &&
-        cachedLayout.widthMeasureMode === widthMeasureMode &&
-        cachedLayout.heightMeasureMode === heightMeasureMode) {
+  function canUseCachedMeasurement(
+      isTextNode,
+      availableWidth,
+      availableHeight,
+      marginRow,
+      marginColumn,
+      widthMeasureMode,
+      heightMeasureMode,
+      cachedLayout) {
+
+    var heightSame =
+      (cachedLayout.heightMeasureMode == CSS_MEASURE_MODE_UNDEFINED && heightMeasureMode == CSS_MEASURE_MODE_UNDEFINED) ||
+        (cachedLayout.heightMeasureMode == heightMeasureMode && cachedLayout.availableHeight == availableHeight);
+
+    var widthSame =
+      (cachedLayout.widthMeasureMode == CSS_MEASURE_MODE_UNDEFINED && widthMeasureMode == CSS_MEASURE_MODE_UNDEFINED) ||
+        (cachedLayout.widthMeasureMode == widthMeasureMode && cachedLayout.availableWidth == availableWidth);
+
+    if (heightSame && widthSame) {
       return true;
     }
 
-    // If the width is an exact match, try a fuzzy match on the height.
-    if (cachedLayout.availableWidth === availableWidth &&
-        cachedLayout.widthMeasureMode === widthMeasureMode &&
-        heightMeasureMode === CSS_MEASURE_MODE_EXACTLY &&
-        availableHeight - marginColumn === cachedLayout.computedHeight) {
+    var heightValid =
+      (cachedLayout.heightMeasureMode == CSS_MEASURE_MODE_UNDEFINED && heightMeasureMode == CSS_MEASURE_MODE_AT_MOST && cachedLayout.computedHeight <= (availableHeight - marginColumn)) ||
+        (heightMeasureMode == CSS_MEASURE_MODE_EXACTLY && cachedLayout.computedHeight == (availableHeight - marginColumn));
+
+    if (widthSame && heightValid) {
+      return true;
+    }
+
+    var widthValid =
+      (cachedLayout.widthMeasureMode == CSS_MEASURE_MODE_UNDEFINED && widthMeasureMode == CSS_MEASURE_MODE_AT_MOST && cachedLayout.computedWidth <= (availableWidth - marginRow)) ||
+        (widthMeasureMode == CSS_MEASURE_MODE_EXACTLY && cachedLayout.computedWidth == (availableWidth - marginRow));
+
+    if (heightSame && widthValid) {
       return true;
     }
 
-    // If the height is an exact match, try a fuzzy match on the width.
-    if (cachedLayout.availableHeight === availableHeight &&
-        cachedLayout.heightMeasureMode === heightMeasureMode &&
-        widthMeasureMode === CSS_MEASURE_MODE_EXACTLY &&
-        availableWidth - marginRow === cachedLayout.computedWidth) {
+    if (heightValid && widthValid) {
       return true;
     }
 
+    // We know this to be text so we can apply some more specialized heuristics.
+    if (isTextNode) {
+      if (widthSame) {
+        if (heightMeasureMode == CSS_MEASURE_MODE_UNDEFINED) {
+          return true;
+        }
+
+        if (heightMeasureMode == CSS_MEASURE_MODE_AT_MOST && cachedLayout.computedHeight < (availableHeight - marginColumn)) {
+          return true;
+        }
+
+        cachedLayout.computedHeight = availableHeight - marginColumn;
```

Comment:
No, the height has not been restricted to be smaller than it was measured so no need to change it